### ISSUE TITLE
Tracking issue for RFC 3621: `derive(CoercePointee)`

### ISSUE THREAD

**Issue by traviscross** [rust-lang/rust#123430]
This is a tracking issue for `#[derive(CoercePointee)]`.

The feature gate for the issue is `#![feature(derive_coerce_pointee)]`.

It was formerly known as:
- `#[derive(UnsizeInner)]`
- `#[derive(SmartPtr)]`
- `#[feature(derive_smart_pointer)]`

### About tracking issues

Tracking issues are used to record the overall progress of implementation.  They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.  A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.  Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Implement the language experiment.
  - https://github.com/rust-lang/rust/pull/125575
  - https://github.com/rust-lang/rust/pull/125048
  - https://github.com/rust-lang/rust/pull/123472
- [x] Accept an RFC.
  - https://github.com/rust-lang/rfcs/pull/3621
- [x] Decide on a final name.
  - https://github.com/rust-lang/rust/issues/129104
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide]).
- [ ] Stabilization PR.
  - https://github.com/rust-lang/rust/pull/133820

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

- [ ] is the derive name good? Should it be more general than "smart pointer"? (Do not discuss on this issue, open a zulip thread or separate GitHub issue)
  - https://github.com/rust-lang/rust/issues/129104

### Related

- https://github.com/rust-lang/rfcs/pull/3621
- https://github.com/rust-lang/rust/pull/125575
- https://github.com/rust-lang/rust/pull/125048
- https://github.com/rust-lang/rust/pull/123472
- https://github.com/rust-lang/rust/pull/127283
- https://github.com/rust-lang/rust/issues/136702
  - fix: https://github.com/rust-lang/rust/pull/136776
- https://github.com/rust-lang/rust/pull/147068
- https://github.com/rust-lang/rust/issues/127323

**Comment by Kixunil** [comment]
It looks like trait bounds are not accounted for: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=879a49c63c2edd517a7594990ebc644d

**Comment by traviscross** [comment]
@Kixunil: If you could, please file that as a separate issue (and reference this tracking issue from it).

**Comment by Darksonn** [comment]
The status is out of date, I'm updating it to reflect that it's blocked on arbitrary self types.

@rustbot label +S-tracking-blocked

**Comment by RalfJung** [comment]
EDIT: Never mind I am stupid.
